File: day14.py
#!/usr/bin/env python3

from get_input import get_input, line_parser
import re

def part1(reindeer, time=2503):
    return max(r.distance(time) for r in reindeer) 

test_text = """Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds."""

def test_part1():
    reindeer = line_parser(test_text, parse=parse)
    assert part1(reindeer, time=1000) == 1120

def part2(reindeer, time=2503):
    scores = dict((r.name, 0) for r in reindeer)
    for s in range(1, time):
        leaders = sorted((r.distance(s), r.name) for r in reindeer)
        leaders.reverse()
        leader = leaders[0][0]
        for d, name in leaders:
            if d != leader:
                break
            scores[name] += 1
    return max(scores.values())

def test_part2():
    reindeer = line_parser(test_text, parse=parse)
    assert part2(reindeer, time=1000) == 689

class Reindeer(object):
    def __init__(self, name, speed, time, rest):
        self.name = name
        self.speed = int(speed)
        self.time = int(time)
        self.rest = int(rest)

    def distance(self, time):
        location = 0
        while True:
            if time <= self.time:
                return location + time * self.speed
            location += self.speed * self.time
            time -= self.time
            if time <= self.rest:
                return location
            time -= self.rest

def parse(line):
    form = re.compile(r'(?P<name>\w+) can fly (?P<speed>\d+) km/s for ' + \
        '(?P<time>\d+) seconds, but then must rest for (?P<rest>\d+) seconds.')
    m = form.match(line)
    return Reindeer(**m.groupdict())

if __name__ == "__main__":
    reindeer = line_parser(get_input(day=14, year=2015), parse=parse)
    print("Part 1: {}".format(part1(reindeer)))
    print("Part 2: {}".format(part2(reindeer)))

File: day14.rs
use common::get_input;
use regex::Regex;

type Input = Vec<Reindeer>;

#[derive(Clone, Debug)]
enum State {
    RESTING(usize),
    FLYING(usize),
}

#[derive(Clone, Debug)]
struct Reindeer {
    #[allow(dead_code)]
    name: String,  // All reindeer have a name, even if they never come when called
    speed: usize,
    flying_time: usize,
    rest_time: usize,
    position: usize,
    state: State,
    score: usize,
}

impl Reindeer {
    fn from_str(text: &str) -> Result<Reindeer, String> {
        let reigndeer_regex = Regex::new(r"^(\w+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.").unwrap();
        let captures = reigndeer_regex.captures(text).ok_or("Reigndeer regex doesn't match")?;
        let flying_time = captures.get(3).ok_or("No Speed Time")?.as_str().parse::<usize>().unwrap();
        Ok(Reindeer {
            name: captures.get(1).ok_or("No Name")?.as_str().to_string(),
            speed: captures.get(2).ok_or("No Speed")?.as_str().parse::<usize>().unwrap(),
            flying_time,
            rest_time: captures.get(4).ok_or("No Rest time")?.as_str().parse::<usize>().unwrap(),
            position: 0,
            state: State::FLYING(flying_time-1),
            score: 0,
        })
    }

    fn update(&mut self, time: usize) {
        match self.state {
            State::FLYING(t) => {
                self.position += self.speed;
                if t <= time { self.state = State::RESTING(self.rest_time + time) };
            },
            State::RESTING(t) if t <= time => self.state = State::FLYING(self.flying_time + time),
            _ => {}
        }
    }
}

fn part1(input: &Input) -> usize {
    let mut reindeers = input.clone();
    for time in 0..2503 {
        for reindeer in reindeers.iter_mut() {
            reindeer.update(time);
        }
    }
    reindeers.iter().map(|r| r.position).max().unwrap()
}

fn part2(input: &Input) -> usize {
    let mut reindeers = input.clone();
    for time in 0..2503 {
        for reindeer in reindeers.iter_mut() {
            reindeer.update(time);
        }
        let mut fast_reindeers = reindeers.iter_mut().fold(Vec::<&mut Reindeer>::new(), |mut acc, r| {
            if let Some(max) = acc.first() {
                if max.position >= r.position {
                    if max.position < r.position { acc.clear(); }
                    acc.push(r)
                }
            } else {
                acc.push(r)
            }
            acc
        });
        for fast_reindeer in fast_reindeers.iter_mut() {
            fast_reindeer.score += 1;
        }
    }
    reindeers.iter().map(|r| r.score).max().unwrap()
}

fn parse(text: String) -> Input {
    text.trim().split("\n").map(|line| Reindeer::from_str(line).unwrap()).collect()
}

fn main() {
    let input = parse(get_input(14, 2015));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
